scripts: enhancements to gen_cfb_font_header

pabigot opened the pull request:
### scripts: gen_cfb_font_header: modify to replicate cfb fonts

The content of subsys/fb/cfb_fonts cannot be replicated by the existing
script due to lack of positioning options and use of a full-color frame
buffer, which affects the generated bitmap.  Switch to the solution used
in the original script, add the required options, and document the
process of regenerating the fonts.

This commit also determines the required bounding box for the glyphs to
be sure that the user-provided value is sufficient to avoid partial
characters.  Ideally the calculated width and height would be used for
font characters, but this would require significant restructuring of the
script to make calculated values available at the point where the
arguments are used to produce output.

### scripts: gen_cfb_font_header: extend to additional representations

The default vertical tiling is designed for displays that are rotated 90
or 270 degrees from normal orientation.  Devices where pixel data
advances first horizontally then vertically requires horizontally-tiled
data.

Similarly when generating upright text on a row-major-order monochrome
display the most significant bit may encode the pixel at the lowest
horizontal position.

Add options to control horizontal vs vertical tiling, and msb vs lsb
pixel order.

The commit also generates the representation of the glyph in comments at
each row.

pabigot committed:
scripts: gen_cfb_font_header: modify to replicate cfb fonts

The content of subsys/fb/cfb_fonts cannot be replicated by the existing
script due to lack of positioning options and use of a full-color frame
buffer, which affects the generated bitmap.  Switch to the solution used
in the original script, add the required options, and document the
process of regenerating the fonts.

This commit also determines the required bounding box for the glyphs to
be sure that the user-provided value is sufficient to avoid partial
characters.  Ideally the calculated width and height would be used for
font characters, but this would require significant restructuring of the
script to make calculated values available at the point where the
arguments are used to produce output.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Thanks to @jfischer-phytec-iot for providing the original script.

This is an initial step towards supporting generation of fonts that use horizontal MSB-first tiling (rather than the current vertical LSB-first tiling).  That's the native format for the displays I'm targeting, but adding it in CFB and display generally will get ugly and I may finish it, so I'm pausing here with something that's at least enough to be able to regenerate the existing character bitmaps.

pabigot commented:
@jfischer-phytec-iot @diegosueiro Apparently nobody owns `subsys/fb`.  All I added was a README, but I shouldn't get ownership because of that.  You're the ones who've done most of the functional coding here; will either of you take it?  Let me know and I'll add a commit.

pabigot commented:
Adding horizontal and MSB was less ugly than I thought, so I've included it, though there are currently no users for it in Zephyr.  This also adds prettier output.

For default vtile / lsb:
```
        /* 125 (}) */
        {
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x30,0x00,0x00,0x30,   /*     XX                      XX   */
                0x30,0x00,0x00,0x30,   /*     XX                      XX   */
                0x70,0x00,0x00,0x30,   /*     XXX                     XX   */
                0x70,0x00,0x00,0x38,   /*     XXX                    XXX   */
                0xe0,0x7f,0xf8,0x1f,   /*      XXXXXXXXXX    XXXXXXXXXX    */
                0xe0,0xff,0xfc,0x1f,   /*      XXXXXXXXXXX  XXXXXXXXXXX    */
                0xc0,0xff,0xfd,0x0f,   /*       XXXXXXXXXXX XXXXXXXXXX     */
                0x00,0x80,0x0f,0x00,   /*                XXXXX             */
                0x00,0x80,0x07,0x00,   /*                XXXX              */
                0x00,0x00,0x03,0x00,   /*                 XX               */
                0x00,0x00,0x03,0x00,   /*                 XX               */
                0x00,0x00,0x03,0x00,   /*                 XX               */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
                0x00,0x00,0x00,0x00,   /*                                  */
```
For htile/msb:

```
	/* 125 (}) */
	{
		0x00,0x00,0x00,   /*                          */
		0x00,0x00,0x00,   /*                          */
		0x00,0x00,0x00,   /*                          */
		0x00,0x00,0x00,   /*                          */
		0x01,0xe0,0x00,   /*        XXXX              */
		0x01,0xf8,0x00,   /*        XXXXXX            */
		0x00,0x7c,0x00,   /*          XXXXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x0f,0x00,   /*             XXXX         */
		0x00,0x07,0xe0,   /*              XXXXXX      */
		0x00,0x03,0xe0,   /*               XXXXX      */
		0x00,0x0f,0x00,   /*             XXXX         */
		0x00,0x1e,0x00,   /*            XXXX          */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x1c,0x00,   /*            XXX           */
		0x00,0x3c,0x00,   /*           XXXX           */
		0x01,0xf8,0x00,   /*        XXXXXX            */
		0x01,0xe0,0x00,   /*        XXXX              */
		0x00,0x00,0x00,   /*                          */
		0x00,0x00,0x00,   /*                          */
```

pabigot committed:
subsys/cfb: improve font structure packing

The specified order of fields wastes space when the cfb_font_caps enum
isn't packed.  Reorder to avoid this behavior.

Also remove the unnecessary array size on the extern symbol declaration,
lest the compiler misinterpret the properties as being zero-length
arrays rather than pointers.  (The idiom is already technically
using undefined behavior since we're relying on the linker rather than
the language to produce an array from the individual declarations.)

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
I now see `htile` should probably be `hpack`; the display and cfb infrastructures use different terms.  Will address after other reviews come in.

jfischer-phytec-iot commented:
>@jfischer-phytec-iot @diegosueiro Apparently nobody owns subsys/fb. All I added was a README, but I shouldn't get ownership because of that. You're the ones who've done most of the functional coding here; will either of you take it? Let me know and I'll add a commit.

@pabigot I think you can add me and @vanwinkeljan?

jfischer-phytec-iot commented on subsys/fb/README_fonts.txt:
```diff
@@ -0,0 +1,19 @@
+The DroidSansMono.ttf font used to generate the cfb_fonts bitmaps can be
```
Maybe it should go into `cfb_fonts.c` file, up to you.

jfischer-no reviewed (approved):
LGTM, thank you

jfischer-phytec-iot commented:
@hakehuang FYI

pabigot commented:
Tagged DNM as there are changes I require that aren't incorporated https://github.com/zephyrproject-rtos/zephyr/pull/17541#issuecomment-511897917.

Also the generated capabilities are wrong.  I'll get to that after I deal with another unblocked PR.

pabigot committed:
CODEOWNERS: add owner for subsys/fb

No owner had been assigned.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
scripts: gen_cfb_font_header: remove guard and make font data private

The output of this script is intended to be put into an implementation
file, where the font data is accessed by index to an array maintained by
the linker script.  There is no need for protection against multiple
includes, and the font data array should not be a global symbol.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
I've updated this with a couple changes:
* a commit that improves packing of the font structure when short enums aren't enabled;
* a commit that refactors in anticipation of msb-first fonts;
* a commit that removes the guard and makes the byte data static

Also I've made the new options match the capabilities name, and corrected the generated capabilities.

This PR got caught in scope creep: the last commit adds support for horizontally-packed and msb-first fonts, which I need, but the driver has not yet been updated to support them.  Doing that would create conflicts with #16331, which probably should be merged first because it's basically ready (modulo my requested changes) and because it provides support for writing to non-byte-aligned offsets which I also need.

This can be merged as-is, or with the last commit which adds unused features removed.  How do you want me to proceed?

jfischer-phytec-iot commented:
I suggest to merge "_with the last commit which adds unused features removed_" and open new PR for cd7d8f8.

pabigot commented:
@jfischer-phytec-iot OK, I've removed hpack and msb-first patches.  Except for the space optimization in the font structure this should be back to about where it was when you reviewed it.  #16331 rebased onto it cleanly.

henrikbrixandersen reviewed (approved):
LGTM, thank you.
